test/apps/iodemo/ucx_wrapper.cc

```diff
@@ -0,0 +1,591 @@
+
+    /* Create worker */
+    ucp_worker_params_t worker_params;
+    worker_params.field_mask  = UCP_WORKER_PARAM_FIELD_THREAD_MODE;
+    worker_params.thread_mode = UCS_THREAD_MODE_SINGLE;
+    status = ucp_worker_create(_context, &worker_params, &_worker);
+    if (status != UCS_OK) {
+        ucp_cleanup(_context);
+        throw UcxError("ucp_worker_create", status);
+    }
+
+    log() << "created worker " << _worker << std::endl;
+
+    post_recv();
+}
+
+UcxContext::~UcxContext()
+{
+    while (!_closing_conns.empty()) {
+        UcxConnection *conn = _closing_conns.front();
+        _closing_conns.pop_front();
+        try {
+            delete conn;
+        } catch (UcxError &e) {
+            debug() << e.what();
+        }
+    }
+
+    while (!_conn_requests.empty()) {
+        assert(_listener);
+        debug() << "reject connection request " << _conn_requests.front();
+        ucp_listener_reject(_listener, _conn_requests.front());
+        _conn_requests.pop_front();
+    }
+
+    // since all connections are handled, close the listener before next
+    // progress to avoid new incoming requests
+    if (_listener) {
+        ucp_listener_destroy(_listener);
+    }
+
+    ucp_request_cancel(_worker, _iomsg_recv_request);
+    try {
+        request_wait("iomsg_recv", _iomsg_recv_request);
+    } catch (const UcxError& e) {
+        debug() << "io msg request is completed with status " << e.status()
+                << std::endl;
+    }
+
+    ucp_worker_destroy(_worker);
+    ucp_cleanup(_context);
+}
+
+void UcxContext::listen(const struct sockaddr* saddr, size_t addrlen)
+{
+    ucp_listener_params_t listener_params;
+
+    listener_params.field_mask         = UCP_LISTENER_PARAM_FIELD_SOCK_ADDR |
+                                         UCP_LISTENER_PARAM_FIELD_CONN_HANDLER;
+    listener_params.sockaddr.addr      = saddr;
+    listener_params.sockaddr.addrlen   = addrlen;
+    listener_params.conn_handler.cb    = connect_callback;
+    listener_params.conn_handler.arg   = reinterpret_cast<void*>(this);
+
+    ucs_status_t status = ucp_listener_create(_worker, &listener_params,
+                                              &_listener);
+    if (status != UCS_OK) {
+        throw UcxError("ucp_listener_create", status);
+    }
+
+    log() << "started listener " << _listener
+          << " on " << sockaddr_str(saddr, addrlen) << std::endl;
+}
+
+UcxConnection* UcxContext::connect(const struct sockaddr* saddr, size_t addrlen)
+{
+    UcxConnection *conn = new UcxConnection(*this, get_next_conn_id(), saddr,
+                                            addrlen);
+    add_connection(conn);
+    return conn;
+}
+
+void UcxContext::disconnect(UcxConnection* conn) _GLIBCXX_NOTHROW
+{
+    debug() << "closing connection " << conn << " with id = " << conn->id()
+            << std::endl;
+    remove_connection(conn);
+    if (!conn->disconnect()) {
+        _closing_conns.push_back(conn);
+    }
+}
```

Comment:
it's confusing that disconnect() would add a "conn" to "closing_conns" and eventually call a virtual disconnect() callback which supposedly overrides the first disconnect()?
what happens if user of ucx_wrapper does not override disconnect()?

Reply:
`conn->disconnect()` is not virtual
`UcxContext::disconnect` is virtual
these classes don't share inheritance tree

> what happens if user of ucx_wrapper does not override disconnect()?

do you mean `UcxContext::disconnect`?
will be called `disconnect()` of base class, user should add handling of its part then call `disconnect()` of base class, this is pretty common usage.

Reply:
i mean UcxContext::disconnect vs. the override in IO demo application

Reply:
> and eventually call a virtual disconnect() callback which supposedly overrides the first disconnect()?

no, these disconnect-methids are not related

> what happens if user of ucx_wrapper does not override disconnect()?

nothing bad, disconnect of base class will be called